ambry-tools/src/main/java/com.github.ambry.tools/perf/rest/NettyPerfClient.java

```diff
@@ -0,0 +1,501 @@
+
+      OptionSet options = parser.parse(args);
+      this.host = options.valueOf(host);
+      logger.info("Host: {}", this.host);
+      this.port = options.valueOf(port);
+      logger.info("Port: {}", this.port);
+      this.path = options.valueOf(path);
+      logger.info("Path: {}", this.path);
+      this.concurrency = options.valueOf(concurrency);
+      logger.info("Concurrency: {}", this.concurrency);
+      this.postBlobTotalSize = options.valueOf(postBlobTotalSize);
+      logger.info("Post blob total size: {}", this.postBlobTotalSize);
+      this.postBlobChunkSize = options.valueOf(postBlobChunkSize);
+      logger.info("Post blob chunk size: {}", this.postBlobChunkSize);
+    }
+  }
+
+  /**
+   * Invokes the {@link NettyPerfClient} with the command line arguments.
+   * @param args command line arguments.
+   */
+  public static void main(String[] args) {
+    try {
+      ClientArgs clientArgs = new ClientArgs(args);
+      final NettyPerfClient nettyPerfClient =
+          new NettyPerfClient(clientArgs.host, clientArgs.port, clientArgs.path, clientArgs.concurrency,
+              clientArgs.postBlobTotalSize, clientArgs.postBlobChunkSize);
+      // attach shutdown handler to catch control-c
+      Runtime.getRuntime().addShutdownHook(new Thread() {
+        public void run() {
+          logger.info("Received shutdown signal. Requesting NettyPerfClient shutdown");
+          nettyPerfClient.shutdown();
+        }
+      });
+      nettyPerfClient.start();
+      nettyPerfClient.awaitShutdown();
+    } catch (Exception e) {
+      logger.error("Exception during execution of NettyPerfClient", e);
+    }
+  }
+
+  /**
+   * Creates an instance of NettyPerfClient
+   * @param host host to contact.
+   * @param port port to contact.
+   * @param path resource path.
+   * @param concurrency number of parallel requests.
+   * @param totalSize the total size in bytes of a blob to be POSTed (-1 if GET is intended).
+   * @param chunkSize size in bytes of each chunk to be POSTed (-1 if GET is intended).
+   */
+  protected NettyPerfClient(String host, int port, String path, int concurrency, long totalSize, int chunkSize) {
+    this.host = host;
+    this.port = port;
+    this.uri = "http://" + host + ":" + port + path;
+    this.concurrency = concurrency;
+    this.totalSize = totalSize;
+    if (chunkSize >= 0) {
+      chunk = new byte[chunkSize];
+      new Random().nextBytes(chunk);
+    } else {
+      chunk = null;
+    }
+    logger.info("Instantiated NettyPerfClient which will interact with host {}, port {}, uri {} with concurrency {}",
+        this.host, this.port, uri, this.concurrency);
+  }
+
+  /**
+   * Starts the NettyPerfClient.
+   * @throws InterruptedException
+   */
+  protected void start()
+      throws InterruptedException {
+    logger.info("Starting NettyPerfClient");
+    reporter.start();
+    group = new NioEventLoopGroup(concurrency);
```

Comment:
It seems concurrency is realized by setting the number of worker threads in `NioEventLoopGroup`. However, how to enforce this is still not very clear to me. For example, if in this line we have `group = new NioEventLoopGroup(2);`, and in ln. 206 we have `for (int i = 0; i < 3; i++) {`, the perf client still runs but 2 and 3 introduce inconsistency. In this case, what is the concurrency then?